crates/lsp/src/handlers_format.rs

```diff
@@ -41,8 +44,148 @@ pub(crate) fn document_formatting(
     Ok(Some(edits))
 }
 
+#[tracing::instrument(level = "info", skip_all)]
+pub(crate) fn document_range_formatting(
+    params: lsp_types::DocumentRangeFormattingParams,
+    state: &WorldState,
+) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
+    let doc = state.get_document(&params.text_document.uri)?;
+
+    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;
+    let range =
+        from_proto::text_range(&doc.line_index.index, params.range, doc.line_index.encoding)?;
+
+    // TODO: Handle FormattingOptions
+    let options = RFormatOptions::default()
+        .with_indent_style(IndentStyle::Space)
+        .with_line_width(line_width);
+
+    let logical_lines = find_deepest_enclosing_logical_lines(doc.parse.syntax(), range);
+    if logical_lines.is_empty() {
+        tracing::warn!("Can't find logical line");
+        return Ok(None);
+    };
+
+    // Find the overall formatting range by concatenating the ranges of the logical lines.
+    // We use the "non-whitespace-range" as that corresponds to what Biome will format.
+    let format_range = logical_lines
+        .iter()
+        .map(|line| text_non_whitespace_range(line))
+        .reduce(|acc, new| acc.cover(new))
+        .expect("`logical_lines` is non-empty");
+
+    // We need to wrap in an `RRoot` otherwise the comments get attached too
+    // deep in the tree. See `CommentsBuilderVisitor` in biome_formatter and the
+    // `is_root` logic. Note that `node` needs to be wrapped in at least two
+    // other nodes in order to fix this problem, and here we have an `RRoot` and
+    // `RExpressionList` that do the job.
+    //
+    // Since we only format logical lines, it is fine to wrap in an expression list.
+    let Some(exprs): Option<Vec<air_r_syntax::AnyRExpression>> = logical_lines
+        .into_iter()
+        .map(|node| air_r_syntax::AnyRExpression::cast(node))
+        .collect()
+    else {
+        tracing::warn!("Can't cast to `AnyRExpression`");
+        return Ok(None);
+    };
+
+    let list = air_r_factory::r_expression_list(exprs);
+    let eof = air_r_syntax::RSyntaxToken::new_detached(RSyntaxKind::EOF, "", vec![], vec![]);
+    let root = air_r_factory::r_root(list, eof).build();
+
+    let format_info = biome_formatter::format_sub_tree(
+        root.syntax(),
+        air_r_formatter::RFormatLanguage::new(options),
+    )?;
+
+    if format_info.range().is_none() {
+        // Happens in edge cases when biome returns a `Printed::new_empty()`
+        return Ok(None);
+    };
+
+    let mut format_text = format_info.into_code();
+
+    // Remove last hard break line from our artifical expression list
+    format_text.pop();
```

Comment:
This comes from `root.rs` when we add the hard line break at the end of the document

Would it be possible to check that the hard line break is there first and only remove it if so? Or erroring if its not there? I worry about dropping a single character if we ever change something in `root.rs`

Reply:
I was thinking the test coverage would prevent that.